### PR TITLE
[QTL] Implement LookupExtractorFactory of namespaced lookup

### PR DESCRIPTION
**Pull request by sirpkt**
This patch adds LookupExtractorFactory of namespaced lookup.
With this, NamespacedExtractor can be used with LookupDimensionSpec.
Done as a first step to merge #2524 with QTL.

Migrated to https://github.com/druid-io/druid/pull/2926

### CODE REVIEW

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    if (manager.scheduleOrUpdateIndividual(extractionNamespace))
+    {
+      // wait until the namespace registration is done
+      String newVersion = manager.getVersion(extractionNamespace.getNamespace());
+      while(newVersion == null || newVersion.equals(oldVersion))
+      {
+        try {
+          Thread.sleep(10);
+        } catch (InterruptedException e) {
+          e.printStackTrace();
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: Should log that it was interrupted (probably as error) and return false.

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+{
+  final ExtractionNamespace extractionNamespace;
+  final NamespaceExtractionCacheManager manager;
+  LookupExtractor lookupExtractor;
+
+  @JsonCreator
+  public NamespaceLookupExtractorFactory(
+      @JsonProperty("extractionNamespace")ExtractionNamespace extractionNamespace,
+      @JacksonInject NamespacedExtractionModule.NamespaceStaticConfiguration namespaceStaticConfiguration,
+      @JacksonInject @Named("dimExtractionNamespace")
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: can these pull from a static final string in the Module?

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    manager.deleteIndividual(extractionNamespace.getNamespace());
+    return true;
+  }
+
+  @Override
+  public boolean replaces(@Nullable LookupExtractorFactory other)
+  {
+    if (other != null)
+    {
+      NamespaceLookupExtractorFactory that = (NamespaceLookupExtractorFactory) other;
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: needs class check

Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: And return true other is not `NamespaceLookupExtractorFactory`

File: extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java [hunk 4]
```diff
@@ -225,38 +225,56 @@ public String apply(
...
+      String namespaceName
+  )
+  {
+    final NamespaceImplData implDatum = implData.get(namespaceName);
+    if (implDatum == null) {
+      // Delete but we don't have it?
+      log.wtf("Asked to delete something I just lost [%s]", namespaceName);
+      return false;
+    }
+    delete(namespaceName);
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java]: suggest return `delete(namespaceName)`

File: extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java [hunk 5]
```diff
@@ -225,38 +225,56 @@ public String apply(
...
+    return true;
+  }
+
+  // return value means actually schedule or not
+  public boolean scheduleOrUpdateIndividual(
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java]: you should be able to leave this as `scheduleOrUpdate` and just have the method parameter be a single `ExtractionNamespace` instead of a `Collection`.

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+import io.druid.query.extraction.namespace.ExtractionNamespace;
+import io.druid.query.lookup.LookupExtractor;
+import io.druid.query.lookup.LookupExtractorFactory;
+import io.druid.server.namespace.NamespacedExtractionModule;
+import io.druid.server.namespace.cache.NamespaceExtractionCacheManager;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+@JsonTypeName("name")
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: suggest making this `namespace` to match `io.druid.query.extraction.NamespacedExtractor`

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+public class NamespaceLookupExtractorFactory implements LookupExtractorFactory
+{
+  final ExtractionNamespace extractionNamespace;
+  final NamespaceExtractionCacheManager manager;
+  LookupExtractor lookupExtractor;
+
+  @JsonCreator
+  public NamespaceLookupExtractorFactory(
+      @JsonProperty("extractionNamespace")ExtractionNamespace extractionNamespace,
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: One limitation here is that ExtractionNamespace has the naming inside the object, and LookupExtractorFactory has the naming outside the object. This means that in the proposed implementation, an identifier must b supplied both outside NamespaceLookupExtractorFactory and inside its ExtractionNamespace parameter. This is certainly not ideal.

Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: I don't have a great solution for this yet, please let me know if you have any ideas.

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java (hunk 1 above)
Comment (sirpkt) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: log added

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java (hunk 2 above)
Comment (sirpkt) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: I'm not sure what you mean.
Is this comment related with string `dimExtractionNamespace`?

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java (hunk 3 above)
Comment (sirpkt) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: Updated based on the comment.
Thank you

File: extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java (hunk 4 above)
Comment (sirpkt) [inline extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java]: updated

File: extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java (hunk 5 above)
Comment (sirpkt) [inline extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java]: updated

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java (hunk 6 above)
Comment (sirpkt) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: updated

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java (hunk 7 above)
Comment (sirpkt) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: I also agree with you that this is not ideal.
And I don't have a good solution yet, too.

If we drop static configuration of namespace lookup completely,
we may auto generate the name of extractionNamespace and use it only internally.

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java (hunk 2 above)
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: Yes. It is starting to get used in a bunch of places and having them all pull from the same static final string helps reduce the chance of fat-fingering a name in the future.

File: extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/NamespacedExtractionModule.java [hunk 8]
```diff
@@ -79,6 +80,7 @@ public void setupModule(SetupContext context)
           {
             context.registerSubtypes(NamespacedExtractor.class);
             context.registerSubtypes(ExtractionNamespace.class);
+            context.registerSubtypes(NamespaceLookupExtractorFactory.class);
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/NamespacedExtractionModule.java]: this is getting to be a long list as well, can it just use `new SimpleModule("DruidNamespacedExtractionModule").registerSubtypes(...)`?

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+      )
+  {
+    this.extractionNamespace = Preconditions.checkNotNull(extractionNamespace,
+        "extractionNamespace should be specified");
+    this.manager = namespaceStaticConfiguration.getManager();
+    this.lookupExtractor = new NamespacedExtractor(fnMaker, reverseFnMaker, extractionNamespace.getNamespace());
+  }
+
+  @Override
+  public boolean start()
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: I feel like what should be happening here is that this object creates the schedule, and an "update" is performed by the thing managing LookupExtractorFactories,

As such, this method could call a `manager.schedule(String id, ExtractionNamespace extractionNamespace, long waitForFirstRun)` kind of method that returns after the extractionNamespace has been populated in the cache at least once, or until waitForFirstRun has timed out waiting.

The id passed does not need to be the ID used in a query, it can simply be a unique ID this NamespaceExtractionFactory uses to reference its part of the cache.

As such this class should have a field which keeps track of its ID used in the cache manager.

On close the delete call to the manager should delete the ID cache.

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java (hunk 7 above)
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: I think having the static config not accessible through the Lookup dimension spec is ok.

**Comment by drcrallen** [comment]
I realized I hadn't written this down yet, but I'm thinking this PR might be able to be the switching point for the extension to allow the recommended lookups to go through the dynamic configuration instead of through static configs for this extension. (mentioned in https://github.com/druid-io/druid/pull/2735#issuecomment-201380789 )

If you think such a task is too far out of scope for this PR, please let me know and we can limit the scope here to a minimal change, and have a potentially more sweeping change come later.

**Comment by sirpkt** [comment]
@drcrallen Ok I'll try to eliminate static configuration of namespace lookup.
As I know, druid-namespace-lookup and druid-kaka-extraction-namespace are the only two extension modules related with lookup, right?

**Comment by drcrallen** [comment]
@sirpkt The only druid.io modules, yes.

**Comment by sirpkt** [comment]
Static config from lookup is eliminated.
Only  `druid.query.extraction.namespace.cache.type` is remained to select type of cache between on/off heap as it is module-wide not per-namespace config.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/extraction/namespace/KafkaExtractionNamespace.java [hunk 10]
```diff
@@ -34,19 +35,19 @@
...
+
   @JsonProperty
-  private final String namespace;
+  private final Properties kafkaProperties;
 
   @JsonCreator
   public KafkaExtractionNamespace(
       @NotNull @JsonProperty(value = "kafkaTopic", required = true) final String kafkaTopic,
-      @NotNull @JsonProperty(value = "namespace", required = true) final String namespace
+      @NotNull @JsonProperty(value = "kafkaProperites", required = true) final Properties kafkaProperties
```
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/extraction/namespace/KafkaExtractionNamespace.java]: Does this need to be required? also required=true is just a hint, you still have to enforce not null checks in constructor to be ACTUALLY required.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java [hunk 11]
```diff
@@ -83,8 +79,7 @@
   // Bindings in KafkaExtractionNamespaceModule
   @Inject
   public KafkaExtractionManager(
-      @Named("namespaceVersionMap") final ConcurrentMap<String, String> namespaceVersionMap,
-      @Named("renameKafkaProperties") final Properties kafkaProperties
+      final Properties kafkaProperties
```
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: This will inject `Properties` from the properties module, not the kafka properties.

File: extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/NamespacedExtractionModule.java [hunk 12]
```diff
@@ -73,14 +75,9 @@
...
         new SimpleModule("DruidNamespacedExtractionModule")
-        {
-          @Override
-          public void setupModule(SetupContext context)
-          {
-            context.registerSubtypes(NamespacedExtractor.class);
-            context.registerSubtypes(ExtractionNamespace.class);
-          }
-        }
+            .registerSubtypes(NamespacedExtractor.class)
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/NamespacedExtractionModule.java]: these can all be on the same method call

File: extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java [hunk 13]
```diff
@@ -199,64 +199,83 @@ public void run()
...
+        } catch (InterruptedException e) {
+          log.error("NamespaceLookupExtractorFactory[%s] - interrupted during start", id);
+          return false;
+        }
+        newVersion = getVersion(id);
       }
-      removeNamespaceLocalMetadata(implDatum);
-      schedule(namespace);
     }
+    return true;
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java]: Should probably return true only if scheduleOrUpdate returns true?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java (hunk 11 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: This is no longer guice injected is it?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java [hunk 14]
```diff
@@ -183,25 +184,28 @@ public void run()
...
               }
               map.put(key, message);
-              namespaceVersionMap.put(namespace, Long.toString(eventCounter.incrementAndGet()));
+              namespaceVersionMap.put(id, Long.toString(eventCounter.incrementAndGet()));
               log.debug("Placed key[%s] val[%s]", key, message);
             }
           }
         }
     );
+    futures.putIfAbsent(id, future);
```
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: Result needs handled for race conditions.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java [hunk 15]
```diff
@@ -213,17 +217,21 @@ public void onFailure(Throwable t)
     );
   }
 
-  @LifecycleStart
+  public void removeListener(final String id)
+  {
+    Future future = futures.remove(id);
```
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: Return value needs check.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionNamespaceFactory.java [hunk 16]
```diff
@@ -38,15 +41,15 @@
  */
 public class KafkaExtractionNamespaceFactory implements ExtractionNamespaceFunctionFactory<KafkaExtractionNamespace>
 {
-  private final KafkaExtractionManager kafkaExtractionManager;
+  private final List<KafkaExtractionManager> kafkaExtractionManagers;
```
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionNamespaceFactory.java]: Why does this require multiple managers now?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionNamespaceModule.java [hunk 17]
```diff
@@ -45,63 +47,26 @@
 public class KafkaExtractionNamespaceModule implements DruidModule
 {
   private static final String PROPERTIES_KEY = "druid.query.rename.kafka.properties";
+  private List<KafkaExtractionManager> kafkaExtractionManagers = Lists.newLinkedList();
```
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionNamespaceModule.java]: Suggest not doing this if at all possible.

File: extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+import io.druid.query.lookup.LookupExtractorFactory;
+import io.druid.server.namespace.NamespacedExtractionModule;
+import io.druid.server.namespace.cache.NamespaceExtractionCacheManager;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+@JsonTypeName("namespace")
+public class NamespaceLookupExtractorFactory implements LookupExtractorFactory
```
Comment (drcrallen) [inline extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java]: Needs full suite of unit tests.

**Comment by drcrallen** [comment]
FYI, I'm taking a crack at the kafka extension because the updates to make it lookup friendly are non-trivial

**Comment by drcrallen** [comment]
@sirpkt Can you revert the kafka namespace extension in this PR?

https://github.com/druid-io/druid/pull/2800

should pretty well handle the kafka portion.

**Comment by drcrallen** [comment]
@sirpkt I fired off https://github.com/sirpkt/druid/pull/1 to try and tinker with some of the functionality. Let me know what you think

**Comment by sirpkt** [comment]
@drcrallen Sorry for the late response and thank you for kind comments.
I think your patch looks great.
On making patch, I wondered why extraction namespace is so complex and Now it becomes much clearer than before.

With your change, however, it is not simple to revert Kafka extraction namespace (just source reverting makes module not working) 
so I think it is better to merge #2800 after it is committed to reduce unnecessary works.

**Comment by drcrallen** [comment]
@sirpkt sure, I think that's reasonable.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java [hunk 19]
```diff
@@ -49,26 +49,22 @@
 import java.util.Map;
 import java.util.Properties;
 import java.util.UUID;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentLinkedQueue;
-import java.util.concurrent.ConcurrentMap;
-import java.util.concurrent.Executors;
+import java.util.concurrent.*;
```
Comment (nishantmonu51) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: could you only import the required classes ?

**Comment by b-slim** [comment]
@sirpkt and @drcrallen what's the status of this PR ?

**Comment by sirpkt** [comment]
@b-slim waiting for the commit of #2800. 
This will be rebased based on it.

**Comment by drcrallen** [comment]
@b-slim excluding the kafka extension, the rest should be about ready.

**Comment by b-slim** [comment]
@drcrallen and @sirpkt can we split this to make review straightforward ?

**Comment by drcrallen** [comment]
@sirpkt Please check out https://github.com/sirpkt/druid/pull/2 which is a merge to master and reverting of the kafka stuff

**Comment by sirpkt** [comment]
merge [sirpkt#2](https://github.com/sirpkt/druid/pull/2) and rebased.

**Comment by drcrallen** [comment]
@b-slim this should be ready for final review (and merge)

**Comment by fjy** [comment]
@b-slim can we finish this up?

**Comment by b-slim** [comment]
@fjy it is in the pipeline...

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.regex.Pattern;
+
+/**
+ *
+ */
+public class KafkaExtractionManager
+{
+  private static final Logger log = new Logger(KafkaExtractionManager.class);
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: can we call all the static variable with upper case please ?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+        return 0L;
+      }
+    }
+  }
+
+  public void addListener(final String id, final KafkaExtractionNamespace kafkaNamespace, final Map<String, String> map)
+  {
+    final String topic = kafkaNamespace.getKafkaTopic();
+    final ListenableFuture<?> future = executorService.submit(
+        new Runnable()
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: should we surround this with try catch ? otherwise any exception will stop the runnable

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+            );
+            List<KafkaStream<String, String>> streams = consumerConnector.createMessageStreamsByFilter(
+                new Whitelist(Pattern.quote(topic)), 1, defaultStringDecoder, defaultStringDecoder
+            );
+
+            if (streams == null || streams.isEmpty()) {
+              throw new IAE("Topic [%s] had no streams", topic);
+            }
+            if (streams.size() > 1) {
+              throw new ISE("Topic [%s] has %d streams! expected 1", topic, streams.size());
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: wondering which case will have more than one stream ?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java (hunk 21 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: This shouldn't have kafka changes, I thought I fixed that. let me see what happened.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+            while (it.hasNext()) {
+              final MessageAndMetadata<String, String> messageAndMetadata = it.next();
+              final String key = messageAndMetadata.key();
+              final String message = messageAndMetadata.message();
+              if (key == null || message == null) {
+                log.error("Bad key/message from topic [%s]: [%s]", topic, messageAndMetadata);
+                continue;
+              }
+              map.put(key, message);
+              namespaceVersionMap.put(id, Long.toString(eventCounter.incrementAndGet()));
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: are we incrementing the version for every key/value read ?

**Comment by b-slim** [comment]
@drcrallen let me know when this is ready.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java (hunk 20 above)
Comment (cheddar) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java]: `log` is `log` across all of the Druid code base...

**Comment by drcrallen** [comment]
@b-slim the non-kafka stuff should be ready

**Comment by drcrallen** [comment]
@sirpkt I'm going to move this PR to a MMX branch. Since most of the reviewers are in my timezone I can have faster modifications while they are awake.

**Comment by drcrallen** [comment]
Moved to https://github.com/druid-io/druid/pull/2926

**Comment by sirpkt** [comment]
@drcrallen OK
